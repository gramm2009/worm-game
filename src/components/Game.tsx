import React, {
  ReactElement,
  useState,
  useRef,
  useCallback,
  useEffect,
} from "react";

interface CellPropsInterface {
  isWorm: boolean;
};

const Cell: React.FC<CellPropsInterface> = ({ isWorm }): ReactElement => {
  return (
    <div className={`game-cell ${isWorm ? "worm" : ""}`}>
      {isWorm ? "WORM" : null}
    </div>
  );
};

const Game = () => {
  const fieldSize = 16;
  let speedRef: React.RefObject<number> = useRef<number>(2000);
  let intervalRef: React.RefObject<NodeJS.Timeout | null> =
    useRef<NodeJS.Timeout | null>(null);

  const [isGameStart, setIsGameStart] = useState<boolean>(false);
  const [score, setScore] = useState<number>(0);
  const [wormPosition, setWormPosition] = useState<number>(
    getRandomNumber(0, fieldSize)
  );

  const renderBoards: () => ReactElement[] = useCallback((): ReactElement[] => {
    const size: number = fieldSize;
    const reactElements: ReactElement[] = [];

    for (let i = 0; i < size; i++) {
      const isWorm = isGameStart && (i === wormPosition ? true : false);
      reactElements.push(<Cell isWorm={isWorm} key={i} />);
    }
    return reactElements;
  }, [isGameStart, wormPosition]);

  function getRandomNumber(min: number, max: number): number {
    return Math.round(Math.random() * (max - min) + min);
  };

  function cellClickHandler(
    e: React.MouseEvent<HTMLDivElement | MouseEvent>
  ): void {
    const target = e.target as HTMLElement;
    const isWorm = Object.values(target.classList).includes("worm");
    if (isWorm) {
      setScore((prev) => prev + 1);
      setWormPosition(getRandomNumber(0, 16));
    };
  };

  function startEndClickHandler(): void {
    setIsGameStart((prev) => !prev);
  };

  const clearIntervslF = useCallback((): void => {
    intervalRef.current && clearInterval(intervalRef.current);
  }, []);

  const setNewIntervalF = useCallback((): void => {
    clearIntervslF();

    intervalRef.current = setInterval(() => {
      setWormPosition(getRandomNumber(0, fieldSize));
    }, speedRef.current);
  }, [clearIntervslF]);


  useEffect(() => {
    if (isGameStart) {
      setNewIntervalF();
    } else {
      clearIntervslF();
    }

    return () => {
      clearIntervslF();
    };
  }, [wormPosition, isGameStart, clearIntervslF, setNewIntervalF]);

  useEffect(() => {
    if (score % 10 === 0) {
      speedRef.current = speedRef.current * 0.8;
    }
  }, [score]);

  return (
    <div className="game">
      <div className="game-title">Click on the worm game</div>
      <div className="game-score">{score}</div>
      <div onClick={cellClickHandler} className="game-field">
        {renderBoards()}
      </div>
      <div>
        <button onClick={startEndClickHandler}>
          {isGameStart ? "PAUSE" : "START"}
        </button>
      </div>
    </div>
  );
};

export default Game;
