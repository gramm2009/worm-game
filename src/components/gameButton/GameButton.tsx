import React from "react";

interface GemeButtonProps {
  startEndClickHandler : ()=>void;
  isGameStart: boolean;
}

const GameButton: React.FC<GemeButtonProps> = ({
  startEndClickHandler,
  isGameStart,
}): React.ReactElement => {
  return (
    <div>
      <button 
        onClick={startEndClickHandler}
        className="game-button"
        role="button"
      >
          {isGameStart ? "PAUSE" : "START"}
      </button>
    </div>
  );
};

export default GameButton;
