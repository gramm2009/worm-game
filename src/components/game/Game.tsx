import React, {
  ReactElement,
  useState,
  useRef,
  useCallback,
  useEffect,
} from "react";
import { getRandomNumber } from '../functions';
import Cell from "../cell/Cell";
import GameButton from "../gameButton/GameButton";

const Game = () => {
  const fieldSize = 16;
  let speedRef: React.RefObject<number> = useRef<number>(2000);
  let intervalRef: React.RefObject<NodeJS.Timeout | null> =
    useRef<NodeJS.Timeout | null>(null);

  const [isGameStart, setIsGameStart] = useState<boolean>(false);
  const [score, setScore] = useState<number>(0);
  const [wormPosition, setWormPosition] = useState<number>(
    getRandomNumber(0, fieldSize)
  );

  const renderBoards: () => ReactElement[] = useCallback((): ReactElement[] => {
    const size: number = fieldSize;
    const reactElements: ReactElement[] = [];

    for (let i = 0; i < size; i++) {
      const isWorm = isGameStart && (i === wormPosition ? true : false);
      reactElements.push(<Cell isWorm={isWorm} key={i} />);
    }
    return reactElements;
  }, [isGameStart, wormPosition]);

  function cellClickHandler(
    e: React.MouseEvent<HTMLDivElement | MouseEvent>
  ): void {
    const target = e.target as HTMLElement;
    const isWorm = Object.values(target.classList).includes("worm");
    if (isWorm) {
      setScore((prev) => prev + 1);
      setWormPosition(getRandomNumber(0, 16));
    }
  }

  const startEndClickHandler = useCallback((): void => {
    setIsGameStart((prev) => !prev);
  }, []);

  const clearIntervslF = useCallback((): void => {
    intervalRef.current && clearInterval(intervalRef.current);
  }, []);

  const setNewIntervalF = useCallback((): void => {
    clearIntervslF();

    intervalRef.current = setInterval(() => {
      setWormPosition(getRandomNumber(0, fieldSize));
    }, speedRef.current);
  }, [clearIntervslF]);

  useEffect(() => {
    if (isGameStart) {
      setNewIntervalF();
    } else {
      clearIntervslF();
    }

    return () => {
      clearIntervslF();
    };
  }, [wormPosition, isGameStart, clearIntervslF, setNewIntervalF]);

  useEffect(() => {
    if (score % 10 === 0) {
      speedRef.current = speedRef.current * 0.8;
    }
  }, [score]);

  return (
    <div className="game" role="game">
      <div className="game-title">Click on the worm game</div>
      <div className="game-score">{score}</div>
      <div onClick={cellClickHandler} className="game-field">
        {renderBoards()}
      </div>
      <GameButton
        startEndClickHandler={startEndClickHandler}
        isGameStart={isGameStart}
      />
    </div>
  );
};

export default Game;
