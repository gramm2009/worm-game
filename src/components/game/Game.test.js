import React from 'react';
import { screen, render, fireEvent, waitFor } from '@testing-library/react';
import Game from './Game';
import GameButton from '../gameButton/GameButton';

describe("Game component render", () => {

  test("Default render", () => {
    render(<Game />);
    const gameElement = screen.getByRole("game");
    expect(gameElement).toBeInTheDocument();
    expect(gameElement).toHaveClass("game");
  });

  test("Render with GameButton", () => {
    render(
      <Game>
        <GameButton />
      </Game>);
    const gameElement = screen.getByRole("game");
    expect(gameElement).toBeInTheDocument();
    expect(gameElement).toHaveClass("game");

    const gameButtonElement = screen.getByRole("button");
    expect(gameButtonElement).toBeInTheDocument();
    expect(gameButtonElement).toHaveClass("game-button");
  });

  test('renders correctly asFragment', () => {
    const { asFragment } = render(<Game />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Integration Game component test", () => {
  test("Click game button", async () => {
    render(<Game />);

    const startButton = screen.getByRole('button', { name: /START/i });
    fireEvent.click(startButton);
    expect(startButton).toHaveTextContent("PAUSE");


    await waitFor(() => {
      const wormCell = screen.getAllByRole("cell").filter(el => el.classList.contains("worm"));
      fireEvent.click(wormCell[0]);
      const scoreElement = screen.getByText(/Score: 1/i);
      expect(scoreElement).toBeInTheDocument();
    });
  });
});
