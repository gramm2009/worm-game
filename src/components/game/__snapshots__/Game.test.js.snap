// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component render renders correctly asFragment 1`] = `
<DocumentFragment>
  <div
    class="game"
    role="game"
  >
    <div
      class="game-title"
    >
      Click on the worm game
    </div>
    <div
      class="game-score"
    >
      0
    </div>
    <div
      class="game-field"
    >
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
      <div
        class="game-cell "
        role="cell"
      />
    </div>
    <div>
      <button
        class="game-button"
        role="button"
      >
        START
      </button>
    </div>
  </div>
</DocumentFragment>
`;
